### DISCUSSION TITLE
Understanding "unit" returned from some functions

### DISCUSSION THREAD

**Discussion by vrtisworks** [savonet/liquidsoap#2286] [Q&A]
I am in the process of upgrading from 1.x to 2.0

This worked in 1.x
localOut = ref output.dummy(blank())
...
some code that eventually sets localOut to a real value
...
source.shutdown(!localOut)

I have tried various combinations of parentheses and :source in different places.  I get various syntax messages.  So, apparently I haven't got a good enough understanding of how this works in 2.0

Can somebody point me in the righth direction or help figure out how to get this working?
Thanks a bunch

**Comment by rrolla** [comment]
I think this is related to 
https://www.liquidsoap.info/doc-dev/migrating.html#returned-types-from-output-operators

**Reply by vrtisworks** [reply]
I read that, and it sounds sort of reasonable.  except the following "works".

localAudio=output.alsa(....,id="localID")
print(localAudio.id())

will print out "localID"
So I think I am missing some basic concept.

**Reply by toots** [reply]
I discovered this `id` method thing recently,. It is kind of an oversight but this is fine.

With `2.x` we decided to switch output returned type to `unit`. This is because outputs should really be leaf nodes in the streaming graph and not be re-used for other purposes.

The attached methods were kept without realizing it, tho this is perfectly fine. In liquisoap `2.x`, any type can have attached methods, for instance an `int` with a `square` method returning its root square.

Lastly, to shutdown an output, I believe you should be able to simply shutdown the source attached to it, for instance:
```
s = single("/path/to/bla.ogg")
output.xxx(..., s)

# Later:
source.shutdown(s)
```
This should also shutdown the output.
